### PR TITLE
Add Policy Recommendation Spark job and image

### PR DESCRIPTION
**Pull request by dreamtalen**
In this PR, we add the policy recommendation spark job and image into the Theia repo. 
Previously policy recommendation spark job has got several reviews on the Antrea repo at https://github.com/antrea-io/antrea/pull/3064. I'm highlighting these changes compared with that closed PR:
1. Add removing auto-generated Pod labels feature to merge recommended policies.
2. Add the capability of recommending the toService ANPs for Pod-to-Svc flows.
3. Make corresponding changes to work with Antctl CLI on the Flow Aggregator side.

Also, this is only the first PR for the Policy Recommendation feature. I will create subsequent PRs including Documentation, Antctl CLI, unit tests, and e2e tests, to complete this new feature.

### CODE REVIEW

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thanks, addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thanks, addressed.

File: build/images/Dockerfile.policy-recommendation.ubuntu
Comment (dreamtalen) [reply build/images/Dockerfile.policy-recommendation.ubuntu:19]: I will add other files like unit tests inside this folder that are unnecessary when building images. Let's keep this as it is for now.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:718]: Sure, add an arg that let users type in a JDBC URL for ClickHouse.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thanks. For `rm_lables` and `to_services`, I was thinking only when users provide `--to_services false` will disable this feature. I rewrite the help message above, please take a second look.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thanks, please take a look again at current list.

File: build/images/Dockerfile.policy-recommendation.ubuntu
Comment (salv-orlando) [inline build/images/Dockerfile.policy-recommendation.ubuntu:1]: todo for a future PR: Consider whether this image can be moved in the antrea dockerhub as we did for the other 3rd party images used by Theia

File: plugins/policy-recommendation/antrea_crd.py
Comment (salv-orlando) [inline plugins/policy-recommendation/antrea_crd.py:19]: With python3 you should no be needing six anymore

File: plugins/policy-recommendation/antrea_crd.py
Comment (salv-orlando) [inline plugins/policy-recommendation/antrea_crd.py:41]: For instance here you can just use .items because you don't have to worry about python2 compatibility.
In case the image defaults to python2, you can change the /usr/bin/python symlink or explicitly trigger the job with python3

File: plugins/policy-recommendation/antrea_crd.py
Comment (salv-orlando) [inline plugins/policy-recommendation/antrea_crd.py:477]: Just a curiosity: it seems you have redefined many objects from the python k8s client because they're used by AntreaNetworkPolicies.

This is fine, but I don't understand why you could not import those from the K8S client library

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [reply plugins/policy-recommendation/policy_recommendation_job.py:718]: pls consider using argparse for options - it's much easier than getopt (this can be done with a follow-up PR)

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:767]: I'd check for -h before entering the loop, a user my specify also other options, we don't want to parse them if -h is specified.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:772]: A very pedant init: you can put the list in a variable, e.g.: valid_types = ["initial", "subsequent"] and then:
```
if arg not in valid_types:
     print("Recommendation type should be in %s" % "or".join(valid_types))
```

Please address this only if you have time, it's not important

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:659]: At some point we need apply pep and pylint to this - there are a lot of violation - but can also be done in a subsequent iteration.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:683]: style nit: avoid commented out lines, if possible.

**Review by salv-orlando — COMMENTED** [review]
Partial review

File: plugins/policy-recommendation/antrea_crd.py
Comment (dreamtalen) [reply plugins/policy-recommendation/antrea_crd.py:19]: Thanks, removed.

File: build/images/Dockerfile.policy-recommendation.ubuntu
Comment (dreamtalen) [reply build/images/Dockerfile.policy-recommendation.ubuntu:1]: Got it, yes we could tag this image and push into our docker hub too.

File: plugins/policy-recommendation/antrea_crd.py
Comment (dreamtalen) [reply plugins/policy-recommendation/antrea_crd.py:41]: Thanks, changed.

File: plugins/policy-recommendation/antrea_crd.py
Comment (dreamtalen) [reply plugins/policy-recommendation/antrea_crd.py:477]: I think I only redefined necessary objects that are not included in the python k8s client or that are different from Antrea's definition. For example, I didn't redefine the `metadata`. But I might be wrong, please let me know if you find any objects that could be reused from the k8s client.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:718]: Thanks for advice, I added a TODO line for now.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:767]: Sounds good to me, thanks!

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:772]: Thanks, addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:659]: Agree, pep/pylint could be done in a Github workflow. I write it down as a TODO too.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:683]: Thanks, addressed.

File: plugins/policy-recommendation/antrea_crd.py
Comment (salv-orlando) [reply plugins/policy-recommendation/antrea_crd.py]: awesome

File: plugins/policy-recommendation/antrea_crd.py
Comment (salv-orlando) [reply plugins/policy-recommendation/antrea_crd.py:477]: Ok, so you redefined only those for which there are changes in Antrea network policy.... that's not a problem. I would have just considered inheriting and replacing attributes, but it's fine also in this way,

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [reply plugins/policy-recommendation/policy_recommendation_job.py:661]: thansk!

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:77]: conversion to list is unnecessary, you can just use labels_dict.keys or

```
labels_dict = dict((x,y) for (x,y) in labels_dict.items() if x not in MEANINGLESS_LABELS)
```

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:112]: This routine might deserve some comment...

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:659]: @yanjunz97  do you think we need to update the clickhouse monitor to periodically clean up recommendation results as well? Do you think perhaps we might need to define an expiration time for results, as perhaps it's not ok to bluntly delete reco results when memory exceeds threshold. 

In any case, I am ok not supporting periodic collelction of old reco results in Theia's first release.

Comment (yanjunz97) [reply plugins/policy-recommendation/policy_recommendation_job.py:659]: As we do not expect the recommendation results may occupy too many spaces, I think expiration time might be more reasonable comparing to cleaning up by the monitor. 

But I'm not sure what expiration time should be chosen. I think a recommendation policy might be useful for a long time. Maybe it is more reasonable to delete them only when users trigger a deletion task from the UI?

File: plugins/policy-recommendation/antrea_crd.py
Comment (yuntanghsu) [inline plugins/policy-recommendation/antrea_crd.py:37]: Is it possible to use inheritance to reuse the func to_dict?

**Comment by salv-orlando** [comment]
It might also make sense of using the python logging library instead of printing to stdout.
It should be fairly easy to introduce logging in this job.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Sounds good to me, changed it to check `flowStartSeconds` instead.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thanks, addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thanks, addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thanks, changed to using dict comprehensions.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:117]: Added.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Make sense to me, marked this as a fatal error, and stopped the spark job immediately.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:523]: In PySpark, I think to achieve a 'namedTuple' like data structure I need to change the current RDD to the Dataframe type. This will involve lots of changes in the computation code. Could we mark this as a TODO for now?

**Comment by dreamtalen** [comment]
> It might also make sense of using the python logging library instead of printing to stdout. It should be fairly easy to introduce logging in this job.

Sure, I added the code to use spark logger to replace the print statement.

File: plugins/policy-recommendation/antrea_crd.py
Comment (dreamtalen) [reply plugins/policy-recommendation/antrea_crd.py:37]: Yes, it's possible, for now, I'm following the code style of https://github.com/kubernetes-client/python/tree/master/kubernetes/client/models. They didn't involve the inheritance for the crds so I think we may don't need to do that too.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (ziyouw) [inline plugins/policy-recommendation/policy_recommendation_job.py:64]: Should we make the log level configrable? It may help for live debug purpose.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (ziyouw) [inline plugins/policy-recommendation/policy_recommendation_job.py:755]: Since we may do not have proper flow records in DB, maybe we need a warn message to indicate this case?

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [reply plugins/policy-recommendation/policy_recommendation_job.py:523]: Surely it can be TODO, Ignore it for this PR.

The namedtuple would actually just be some syntactic sugar, where you access item in a tuple as if you were accessing an object.

I would not think this requires using a DataFrame, but if that's the case, it's surely not worth the effort.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [reply plugins/policy-recommendation/policy_recommendation_job.py:668]: Thanks @yanjunz97, that's valuable feedback. Nothing we need to address here, but we surely need a mechanism to handle lifecycle of policy recommendation results.

**Review by salv-orlando — COMMENTED** [review]
Code looks pretty much good to me.
There are a few pending questions from Ziyou, waiting on approval.

File: plugins/policy-recommendation/antrea_crd.py
Comment (annakhm) [inline plugins/policy-recommendation/antrea_crd.py:220]: ist->list

File: plugins/policy-recommendation/antrea_crd.py
Comment (annakhm) [inline plugins/policy-recommendation/antrea_crd.py:105]: nit: kubernete->kubernetes here and below

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:64]: Thank Salvatore help me answer this question. 
Yes, I hard-coded log level to info because I only added logs at info, warning, and error level. Also tried changing to "debug" mode and I could see lots of debug logs automatically generated by Spark.

File: plugins/policy-recommendation/antrea_crd.py
Comment (dreamtalen) [reply plugins/policy-recommendation/antrea_crd.py:335]: Thanks, addressed.

File: plugins/policy-recommendation/antrea_crd.py
Comment (dreamtalen) [reply plugins/policy-recommendation/antrea_crd.py:105]: Thanks, addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py:233]: Thanks, addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thank Jianjun, addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Addressed.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (salv-orlando) [inline plugins/policy-recommendation/policy_recommendation_job.py:652]: nit (perhaps to be addressed in a future PR): id is a reserved keyword in python. In this case it will be correctly interpreted, but using it is a risk for a maintainability perspective (e.g.: we change the param name to something else, no error is thrown when running the code, but "if not id" will always be false!)

**Review by salv-orlando — APPROVED** [review]
I think code LGTM. Hope issues with logging have been sorted out.

File: plugins/policy-recommendation/policy_recommendation_job.py
Comment (dreamtalen) [reply plugins/policy-recommendation/policy_recommendation_job.py]: Thank Salvatore, that's a fair concern. I rename this parameter to `recommendation_id_input` instead.

**Review by annakhm — APPROVED** [review]

**Pull request merged**
